tests/testsuite/artifact_dep.rs

```diff
@@ -1568,12 +1568,174 @@ fn artifact_dep_target_specified() {
+        .with_stdout_data(str![[r#"
+foo v0.0.0 ([ROOT]/foo)
+└── bindep v0.0.0 ([ROOT]/foo/bindep)
+
+"#]])
+        .with_status(0)
+        .run();
+}
+
+/// From issue #10061
+/// The case where:
+/// *   artifact dep is { target = <specified> }
+/// *   dependency of that artifact dependency specifies the same target
+/// *   the target is not activated.
+#[cargo_test]
+fn dep_of_artifact_dep_same_target_specified() {
+    if cross_compile::disabled() {
+        return;
+    }
+    let target = cross_compile::alternate();
+    let p = project()
+        .file(
+            "Cargo.toml",
+            &format!(
+                r#"
+                    [package]
+                    name = "foo"
+                    version = "0.1.0"
+                    edition = "2015"
+                    resolver = "2"
+
+                    [dependencies]
+                    bar = {{ path = "bar", artifact = "bin", target = "{target}" }}
+                "#,
+            ),
+        )
+        .file("src/lib.rs", "")
+        .file(
+            "bar/Cargo.toml",
+            &format!(
+                r#"
+                    [package]
+                    name = "bar"
+                    version = "0.1.0"
+
+                    [target.{target}.dependencies]
+                    baz = {{ path = "../baz" }}
+                "#,
+            ),
+        )
+        .file("bar/src/lib.rs", "")
+        .file(
+            "baz/Cargo.toml",
+            r#"
+                [package]
+                name = "baz"
+                version = "0.1.0"
+
+            "#,
+        )
+        .file("baz/src/lib.rs", "")
+        .build();
+
+    // TODO This command currently fails due to a bug in cargo but it should be fixed so that it succeeds in the future.
+    p.cargo("check -Z bindeps")
+        .masquerade_as_nightly_cargo(&["bindeps"])
+        .with_stderr_data(str![[r#"
+[LOCKING] 2 packages to latest compatible versions
+[ERROR] dependency `bar` in package `foo` requires a `bin` artifact to be present.
+
+"#]])
+        .with_status(101)
+        .run();
+
+    // TODO This command currently fails due to a bug in cargo but it should be fixed so that it succeeds in the future.
+    p.cargo("tree -Z bindeps")
+        .masquerade_as_nightly_cargo(&["bindeps"])
+        .with_stderr_data(
+            r#"...
+no entry found for key
+...
+"#,
+        )
+        .with_status(101)
+        .run();
+}
+
+#[cargo_test]
+fn check_bindep_depending_on_build_only_dep_with_optional_dep() {
+    if cross_compile::disabled() {
+        return;
+    }
+    // i686-unknown-linux-gnu did not trigger a panic so we need to manually specify wasm32-unknown-unknown instead of relying on cross_compile::alternate()
+    // This is possibly because the `cfg(unix)` later on needs to evaluate to false on the cross compile but true on the host.
+    let target = "wasm32-unknown-unknown";
```

Comment:
Isn't it fine that this test fails, since `wasm32-unknown-unknown` isn't installed on the ?

After adding the target on my own computer (`rustup target add wasm32-unknown-unknown`), the test passes. Therefore I will delete this unnecessary test.

Reply:
I feel that we don't really need this test. My instinct tells me that they share the same bug.